File: z1/python/eventify_client.py
import sys
import asyncio
import grpc
import time
import eventify_pb2 as eventify
import eventify_pb2_grpc as eventify_grpc

async def subscribe_to_type(stub, type_name):
    subscription_string = "Subscribed: %s" % type_name
    print(subscription_string)
    print("-" * len(subscription_string))

    reconnect_interval = 1
    request = eventify.SubRequest(type=string_to_event_type(type_name))

    while True:
        try:
            events = stub.Subscribe(request)

            for event in events:
                print("New %s event: %s" % (type_name, event))
                reconnect_interval = 1  # Got connection, reset timer
        except grpc.RpcError as e:
            print("Connectivity issue: %s" % e.details())
            time.sleep(reconnect_interval)
            reconnect_interval = increase_interval(reconnect_interval)

def string_to_event_type(type_name):
    event_types = {
        "CONCERT": eventify.EventType.CONCERT,
        "CONFERENCE": eventify.EventType.CONFERENCE,
        "FESTIVAL": eventify.EventType.FESTIVAL
    }

    return event_types[type_name.upper()]

def increase_interval(interval):
    new_interval = interval * 2
    if new_interval > 30:
        new_interval = 30

    return new_interval

async def main():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = eventify_grpc.EventifyStub(channel)
        coroutines = []
        for event_type in sys.argv[1:]:
            coroutines.append(subscribe_to_type(stub, event_type))

        await asyncio.gather(*coroutines)

if __name__ == '__main__':
    asyncio.run(main())

File: z1/ruby/eventify_client.rb
this_dir = File.expand_path(__dir__)
lib_dir = File.join(this_dir, 'lib')
$LOAD_PATH.unshift(lib_dir) unless $LOAD_PATH.include?(lib_dir)

require 'eventify_services_pb'
require 'grpc'

include Eventify

def subscribe_for_type(stub, type_name)
  subscription_string = "Subscribed: #{type_name}"
  puts(subscription_string)
  puts('-' * subscription_string.length)

  request = SubRequest.new(
    type: string_to_event_type(type_name)
  )

  events = stub.subscribe(request)
  events.each do |event|
    puts("New #{type_name} event: #{event}")
  end
end

def string_to_event_type(name)
  case name.upcase
  when 'CONCERT'
    EventType::CONCERT
  when 'CONFERENCE'
    EventType::CONFERENCE
  when 'FESTIVAL'
    EventType::FESTIVAL
  else
    raise ArgumentError, "Unrecognized event type: #{name}"
  end
end

def main
  stub = Eventify::Eventify::Stub.new('localhost:50051', :this_channel_is_insecure)

  if ARGV.empty?
    puts('Supply at least one event type')
    exit
  end

  ARGV.each do |event_type|
    begin
      Thread.new do
        subscribe_for_type(stub, event_type)
      end
    rescue ArgumentError => e
      puts(e)
    end
  end

  loop do
    STDIN.gets
  end
end

begin
  main
rescue Interrupt => _e
  exit
end
